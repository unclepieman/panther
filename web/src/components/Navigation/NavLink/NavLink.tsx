/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Box, Icon, IconProps } from 'pouncejs';
import React from 'react';
import useRouter from 'Hooks/useRouter';
import { addTrailingSlash, getPathnameFromURI } from 'Helpers/utils';
import { Link as RRLink } from 'react-router-dom';

type NavLinkProps = {
  icon: IconProps['type'];
  label: string;
  to: string;
  isSecondary?: boolean;
};

const NavLink: React.FC<NavLinkProps> = ({ icon, label, to, isSecondary }) => {
  const { location } = useRouter();
  const pathname = addTrailingSlash(location.pathname);
  const destination = addTrailingSlash(getPathnameFromURI(to));
  const isActive = pathname.startsWith(destination);

  const activeColor = isSecondary || isActive ? 'blue-400' : 'navyblue-500';
  const backgroundColor = isActive ? 'blue-400' : 'transparent';

  return (
    <Box as={RRLink} display="block" to={to} aria-current={isActive ? 'page' : undefined}>
      <Box
        borderRadius="small"
        color="gray-50"
        fontSize={isSecondary ? 'small-medium' : 'medium'}
        display="flex"
        alignItems="center"
        px={isSecondary ? 2 : 4}
        py={3}
        backgroundColor={backgroundColor}
        _hover={{
          backgroundColor: activeColor,
        }}
        _focus={{
          backgroundColor: activeColor,
        }}
        transition="background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms"
        truncated
      >
        <Icon type={icon} size={isSecondary ? 'small' : 'medium'} mr={isSecondary ? 4 : 3} />
        <Box>{label}</Box>
      </Box>
    </Box>
  );
};

export default NavLink;
