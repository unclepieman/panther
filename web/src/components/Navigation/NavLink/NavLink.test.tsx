/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render, screen } from 'test-utils';
import NavLink from './index';

describe('NavLink', () => {
  const navLinkDisplayName = 'alerts';
  const getNavLink = () =>
    screen.getByText((content, element) => {
      return element.tagName.toLowerCase() === 'a' && element.textContent === navLinkDisplayName;
    });

  it('could render a secondary navigation link', () => {
    const { container } = render(
      <NavLink to="/something/" label={navLinkDisplayName} icon="list" isSecondary />,
      {
        initialRoute: '/something/particular/',
      }
    );

    expect(container).toMatchSnapshot();
  });

  it('matches a URI regardless of hashes or query params', () => {
    render(<NavLink to="/something/" label={navLinkDisplayName} icon="list" />, {
      initialRoute: '/something/#whatever?q=anything',
    });

    expect(getNavLink()).toHaveAttribute('aria-current', 'page');
  });

  it('matches a URI regardless of trailing slashes', () => {
    render(<NavLink to="/whatever" label={navLinkDisplayName} icon="list" />, {
      initialRoute: `/whatever/`,
    });

    expect(getNavLink()).toHaveAttribute('aria-current', 'page');
  });

  it('matches children URIs', () => {
    render(<NavLink to="/something/" label={navLinkDisplayName} icon="list" />, {
      initialRoute: '/something/particular/',
    });

    expect(getNavLink()).toHaveAttribute('aria-current', 'page');
  });

  it('ignores unrelated URIs', () => {
    render(<NavLink to="/something/" label={navLinkDisplayName} icon="list" />, {
      initialRoute: '/something-else/',
    });

    expect(getNavLink()).not.toHaveAttribute('aria-current', 'page');
  });
});
