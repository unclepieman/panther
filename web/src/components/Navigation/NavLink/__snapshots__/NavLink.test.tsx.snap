// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLink could render a secondary navigation link 1`] = `
.panther-3 {
  display: block;
}

.panther-2 {
  border-radius: 2px;
  color: #f6f6f6;
  font-size: 0.813rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 12px;
  padding-bottom: 12px;
  background-color: #0081c5;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.panther-2:hover {
  background-color: #0081c5;
}

.panther-2:focus {
  background-color: #0081c5;
}

.panther-0 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  color: currentColor;
  margin-right: 16px;
}



<div>
  <header
    id="main-header"
  />
  <a
    aria-current="page"
    class="panther-3"
    href="/something/"
  >
    <div
      class="panther-2"
    >
      <svg
        class="panther-0"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 7H17V9H11V7ZM11 11H17V13H11V11ZM11 15H17V17H11V15ZM7 7H9V9H7V7ZM7 11H9V13H7V11ZM7 15H9V17H7V15ZM3 3V21H21V3H3ZM19 19H5V5H19V19Z"
        />
      </svg>
      <div
        class="panther-1"
      >
        alerts
      </div>
    </div>
  </a>
  <footer
    id="footer"
  />
</div>
`;
